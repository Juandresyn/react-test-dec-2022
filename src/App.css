#root {
  width: 1100px;
  margin: 0 auto;
  padding: 2rem;
  text-align: center;
  --nav-bar-width: 230px;
}

.logo {
  height: 6em;
  padding: 1.5em;
  will-change: filter;
}
.logo:hover {
  filter: drop-shadow(0 0 2em #646cffaa);
}
.logo.react:hover {
  filter: drop-shadow(0 0 2em #61dafbaa);
}

@media (prefers-reduced-motion: no-preference) {
  a:nth-of-type(2) .logo {
    animation: logo-spin infinite 20s linear;
  }
}

.card {
  padding: 2em;
}

.main {
  display: flex;
}

.toolbar {
  width: var(--nav-bar-width);
  text-align: left;
  padding-right: 30px;
  margin-top: 120px;
}

.menu {
  padding: 0;
  margin: 0;
  list-style: none;
  text-transform: uppercase;
}

.menu li {
  padding: 10px 0;
}

.menu a {
  font-weight: bold;
}

.content {
  width: calc(100% - var(--nav-bar-width));
}