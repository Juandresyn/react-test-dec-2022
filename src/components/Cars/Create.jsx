import { useState } from 'react';
import PropTypes from 'prop-types';

function CreateReservation({ submit }) {
  const [id, setId] = useState(null);
  const [maker, setMaker] = useState(null);
  const [model, setModel] = useState(null);
  const [ref, setRef] = useState(null);
  const [color, setColor] = useState(null);
  const [milage, setMilage] = useState(null);

  /**
   * Default object representation
   */
  const responseBody = {
    id: '',
    maker: '',
    model: 0,
    ref: '',
    color: '',
    milage: 0,
  };

  /**
   * Handle form submit
   */
  const handleSubmit = (event) => {
    event.preventDefault();

    responseBody.id = id;
    responseBody.maker = maker;
    responseBody.model = model;
    responseBody.ref = ref;
    responseBody.color = color;
    responseBody.milage = milage;

    submit(responseBody);
  };

  const inputChangeHandler = (setFunction, event) => setFunction(event.target.value);

  return (
    <form className="form CreateReservation" onSubmit={handleSubmit}>
      <div className="form__wrapper">
        <input role="textbox" type="text" placeholder="License" name="id" onChange={(e) => inputChangeHandler(setId, e)} />
        <input role="textbox" type="text" placeholder="Maker" name="maker" onChange={(e) => inputChangeHandler(setMaker, e)} />
        <input role="textbox" type="text" placeholder="Model" name="Model" onChange={(e) => inputChangeHandler(setModel, e)} />
        <input role="textbox" type="text" placeholder="Ref" name="Ref" onChange={(e) => inputChangeHandler(setRef, e)} />
        <input role="textbox" type="text" placeholder="Color" name="Color" onChange={(e) => inputChangeHandler(setColor, e)} />
        <input role="textbox" type="text" placeholder="Milage" name="Milage" onChange={(e) => inputChangeHandler(setMilage, e)} />
      </div>
      <button role="button">Create Car</button>
    </form>
  );
}

CreateReservation.propTypes = {
  submit: PropTypes.func.isRequired,
};

export default CreateReservation;
