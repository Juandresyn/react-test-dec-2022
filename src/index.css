:root {
  font-family: Inter, Avenir, Helvetica, Arial, sans-serif;
  font-size: 16px;
  line-height: 24px;
  font-weight: 400;

  color-scheme: light dark;
  color: rgba(255, 255, 255, 0.87);
  background-color: #242424;

  font-synthesis: none;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-text-size-adjust: 100%;
}

a {
  font-weight: 500;
  color: #646cff;
  text-decoration: inherit;
}
a:hover {
  color: #535bf2;
}

body {
  margin: 0;
  display: flex;
  min-width: 320px;
  min-height: 100vh;
}

h1 {
  font-size: 3.2em;
  line-height: 1.1;
}

button {
  border-radius: 8px;
  border: 1px solid transparent;
  padding: 0.6em 1.2em;
  font-size: 1em;
  font-weight: 500;
  font-family: inherit;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.25s;
}
button:hover {
  border-color: #646cff;
}
button:focus,
button:focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

.button--wide {
  width: 100%;
  margin-bottom: 30px;
}

.form {
  display: flex;
  flex-direction: column;
  margin-bottom: 35px;
}

.form__wrapper {
  display: flex;
  flex-direction: column;
}

.form__group {
  display: flex;
  justify-content: space-between;
}

.form__group > * {
  width: 49% !important;
}

.form__wrapper input,
.form__wrapper select,
.form__wrapper textarea {
  width: 100%;
  padding: 15px 10px;
  border: 1px solid #cdcdcd;
  border-radius: 3px;
  margin-bottom: 14px;
  box-sizing: border-box;
}

.title__main {
  text-align: left;
}

@media (prefers-color-scheme: light) {
  :root {
    color: #213547;
    background-color: #ffffff;
  }
  a:hover {
    color: #747bff;
  }
  button {
    background-color: #f9f9f9;
  }
}
