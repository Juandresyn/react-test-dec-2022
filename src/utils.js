export const transformDate = (date, showTime = true) => {
    const year = date.getFullYear();
    const month = date.getMonth() + 1;
    const day = date.getDate();
    const hour = date.getHours();
    const minutes = date.getMinutes();

    const time = `${hour}:${minutes}:00`;

    return `${year}-${month}-${day} ${showTime ? time : ''}`
};

export const getAge = (dob) => {
    const ageDifMs = Date.now() - new Date(dob).getTime();
    const ageDate = new Date(ageDifMs);

    return Math.abs(ageDate.getUTCFullYear() - 1970);
}

export const users = [
    {
        id: '23456789',
        name: 'Pepito',
        lastName: 'Perez',
        dob: '1973-06-12'
    },
    {
        id: '12345678',
        name: 'Juan',
        lastName: 'Yepes',
        dob: '1992-02-16'
    }
];

export const cars = [
    {
      id: "NOF123",
      maker: 'mercedes-benz',
      model: 2023,
      ref: 'GLC300',
      color: 'blue',
      milage: 31278
    },
    {
      id: "IHJ889",
      maker: 'audi',
      model: 2020,
      ref: 'Q5',
      color: 'white',
      milage: 45123
    },
    {
      id: "UUE654",
      maker: 'mazda',
      model: 2021,
      ref: 'CX-9',
      color: 'silver',
      milage: 55687
    }
];

export const reservations = [
    {
      id: "rid-20221210-000001",
      user: {
        id: '12345678',
        name: 'Juan',
        lastName: 'Yepes',
        age: '30'
      },
      car: {
        id: "NOF123",
        maker: 'mercedes-benz',
        model: 2023,
        ref: 'GLC300',
        color: 'blue',
        milage: 31278
      },
      from: '2022-12-10 10:00:00',
      to: '2022-12-20 11:00:00',
      notes: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed volutpat, nibh nec egestas sodales, justo nulla sodales dolor.'
    },
    {
      id: "rid-20221210-000002",
      user: {
        id: '23456789',
        name: 'Pepito',
        lastName: 'Perez',
        age: '45'
      },
      car: {
        id: "IHJ889",
        maker: 'audi',
        model: 2020,
        ref: 'Q5',
        color: 'white',
        milage: 45123
      },
      from: '2022-12-10 10:30:00',
      to: '2022-12-20 11:10:00',
      notes: ''
    },
];