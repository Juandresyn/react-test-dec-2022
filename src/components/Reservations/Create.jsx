import { useState } from 'react';
import DatePicker from 'react-datepicker';
import "react-datepicker/dist/react-datepicker.css";
import { transformDate } from '../../utils';

function CreateReservation({ submit }) {
  const [user, setUser] = useState(null);
  const [car, setCar] = useState(null);
  const [notes, setNotes] = useState(null);
  const [startDate, setStartDate] = useState(null);
  const [endDate, setEndDate] = useState();

  /**
   * Default object representation
   */
  const responseBody = {
    userId: '',
    carId: '',
    from: null,
    to: null,
    notes: ''
  };

  /**
   * Handle form submit
   */
  const handleSubmit = (event) => {
    event.preventDefault();
    responseBody.userId = parseInt(user);
    responseBody.carId = car;
    responseBody.from = transformDate(startDate);
    responseBody.to = transformDate(endDate);
    responseBody.notes = notes;

    submit(responseBody);
  };

  /**
   * Handle form when and input changes its value.
   */
  const inputChangeHandler = (setFunction, event) => setFunction(event.target.value);

  return (
    <form className="form CreateReservation" onSubmit={handleSubmit}>
      <div className="form__wrapper">
          <input type="text" placeholder="User" name="user" onChange={(e)=>inputChangeHandler(setUser, e)} />
          <input type="text" placeholder="Car" name="car" onChange={(e)=>inputChangeHandler(setCar, e)} />
          <div className="form__group">
            <DatePicker
              selected={startDate}
              showTimeSelect
              onChange={(date) => setStartDate(date)}
              minDate={new Date()}
              selectsStart
              showDisabledMonthNavigation
              endDate={endDate}
              placeholderText="From"
            />
            <DatePicker
              selected={endDate}
              showTimeSelect
              onChange={(date) => setEndDate(date)}
              selectsEnd
              startDate={startDate}
              endDate={endDate}
              minDate={startDate}
              placeholderText="To"
            />
          </div>
          <textarea name="notes" id="" cols="30" rows="5" placeholder="Notes" onChange={(e)=>inputChangeHandler(setNotes, e)}></textarea>
      </div>
      <button>Create Reservation</button>
    </form>
  )
}

export default CreateReservation;
