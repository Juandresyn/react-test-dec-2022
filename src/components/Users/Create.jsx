import { useState } from 'react';
import DatePicker from 'react-datepicker';
import "react-datepicker/dist/react-datepicker.css";
import { transformDate } from '../../utils';

function CreateReservation({ submit }) {
  const [id, setId] = useState(null);
  const [name, setName] = useState(null);
  const [lastname, setLastname] = useState(null);
  const [dob, setDob] = useState(null);

  const responseBody = {
    id: '',
    name: '',
    lastname: 0,
    dob: null
  };

  const handleSubmit = (event) => {
    event.preventDefault();

    responseBody.id = id;
    responseBody.name = name;
    responseBody.lastname = lastname;
    responseBody.dob = transformDate(dob, false);

    submit(responseBody);
  };

  const inputChangeHandler = (setFunction, event) => setFunction(event.target.value);

  return (
    <form className="form CreateReservation" onSubmit={handleSubmit}>
      <div className="form__wrapper">
          <input type="number" placeholder="ID" name="id" onChange={(e)=>inputChangeHandler(setId, e)} />
          <input type="text" placeholder="Name" name="maker" onChange={(e)=>inputChangeHandler(setName, e)} />
          <input type="text" placeholder="Last Name" name="Model" onChange={(e)=>inputChangeHandler(setLastname, e)} />
          <DatePicker
              selected={dob}
              onChange={(date) => setDob(date)}
              peekNextMonth
              showMonthDropdown
              showYearDropdown
              dropdownMode="select"
              placeholderText="DoB"
              maxDate={Date.now()}
            />
      </div>
      <button>Create User</button>
    </form>
  )
}

export default CreateReservation;
