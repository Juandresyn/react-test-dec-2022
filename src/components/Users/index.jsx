import { useState, useCallback, useEffect } from 'react';
import useFetch, { Provider } from 'use-http';
import Table from '../Table';
import Create from './Create';
import { getAge } from '../../utils.js';
import { ENDPOINT_BASE_URL, USER_EP } from '../../config/api';

const headers = ['id', 'cedula', 'name', 'lastname', 'age'];
const transformRows = (row) => (row && row.name ? {
  id: row.client,
  cedula: row.id,
  name: row.name,
  lastname: row.lastname,
  dob: getAge(row.dob)
} : {});

function Users() {
  const [showForm, setShowForm] = useState(false);
  const [userList, setUserList] = useState([]);

  /**
   * Initialize useFetch
   */
  const { get, post, del, response, loading, error } = useFetch(ENDPOINT_BASE_URL, { data: [] });

  /**
   * Load users from endpoint
   */
  const loadInitialUsers = useCallback(async () => {
    const initialUsers = await get(USER_EP);
    if (response.ok) setUserList(initialUsers.data)
  }, [get, response]);

  /**
   * componentDidMount
   */
  useEffect(() => { loadInitialUsers() }, [loadInitialUsers]);

  /**
   * send new user data to the server for creation
   */
  const postNewUser = useCallback(async (data) => {
    if (!!data && !data.id) return;
    const newUser = await post(USER_EP, data);

    if (response.ok) {
      setUserList(items => [newUser.data, ...items]);
      setShowForm(showForm => false);
    }
  }, [post, response, userList]);

  /**
   * Ask server to delete an existing user
   */
  const deleteUser = useCallback(async (id) => {
    await del(`${ USER_EP }/${id}`);

    if (response.ok) {
      setUserList(users => users.filter(c => c.id === id))
    }
  }, [del, response, userList]);

  /**
   * Handle for when <Create> is submitted
   */
  const handleSubmit = (data) => {
    postNewUser(data);
  };

  return (
    <div className="Users">
      <h1 className="title__main">Manage Users</h1>

      <button role="button" className="button--wide" onClick={() => setShowForm((showForm) => !showForm)}>{ showForm ? 'Hide Form' : 'Create User' }</button>

      { showForm ? <Create submit={ (data) => handleSubmit(data) }></Create> : null }

      { !loading ? `${ userList.length } users found` : '' }
      <Table
        headers={headers}
        items={userList.map(transformRows)}
        options={({
          remove: deleteUser
        })}></Table>
    </div>
  )
}

export default Users;
