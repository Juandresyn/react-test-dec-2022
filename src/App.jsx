import { useState } from 'react';
import {
  createBrowserRouter,
  RouterProvider,
} from "react-router-dom";
import reactLogo from './assets/react.svg';
import Cars from './components/Cars';
import Users from './components/Users';
import Reservations from './components/Reservations';
import Logo from './assets/logo.png';
import './App.css';

const router = createBrowserRouter([
  {
    path: "/",
    element: <h1>Rent-a-Car</h1>,
  },
  {
    path: "/manage/cars",
    element: <Cars></Cars>,
  },
  {
    path: "/manage/users",
    element: <Users></Users>,
  },
  {
    path: "/manage/reservations",
    element: <Reservations></Reservations>,
  },
]);

function App() {
  const [count, setCount] = useState(0)

  return (
    <div className="App">
      <div>
        <a href="/">
          <img src={ Logo } className="logo" alt="Company Logo" />
        </a>
      </div>
      <main className="main">
        <nav className="toolbar">
          <ul className="menu">
            <li>
              <a href="/manage/cars">Manage Cars</a>
            </li>
            <li>
              <a href="/manage/users">Manage Users</a>
            </li>
            <li>
              <a href="/manage/reservations">Manage Reservations</a>
            </li>
          </ul>
        </nav>
        <div className="content">
          <RouterProvider router={router} />
        </div>
      </main>
    </div>
  )
}

export default App;
