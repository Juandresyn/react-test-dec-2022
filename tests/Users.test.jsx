
import { render, screen, waitFor, fireEvent, userEvent, act } from "@testing-library/react";
import { useFetch } from 'use-http';
import Users from '../src/components/Users';
import Create from '../src/components/Users/Create';
import { users } from '../src/mockData';
import * as utils from '../src/utils';

jest.mock('use-http');

jest.mock('react-datepicker/dist/react-datepicker.css', () => ({
  __esModule: true,
  default: {
    class1: 'class1'
  },
}));

describe('Test <Users /> component', () => {
  const mockFetch = { useFetch };
  useFetch.mockReturnValue({
      get: () => ({ data: users }),
      post: () => null,
      del: () => null,
      response: { ok: true },
      loading: false
    });

  test('Should call endpoint', async () => {
    let Render;

    act(() => Render = render(<Users></Users>));

    await waitFor(() => {
      const plainHTML = Render.asFragment();
      expect(plainHTML).toMatchSnapshot();


      const usersCount = screen.getByText(`${ users.length } users found`);
      expect(usersCount).toBeTruthy();
    });
  });

  test('Should open <Create /> component', async () => {
    let Render;

    act(() => Render = render(<Users></Users>));

    await waitFor(() => {
      fireEvent.click(Render.container.querySelector('[role="button"]'));

      const button = screen.getByRole('button', {
        name: /hide form/i
      });

      expect(button).toBeTruthy();
    });
  });
});

describe('Test <Create /> component', () => {
  test('<Create /> component should pass data on submit', async () => {
    const createDataMock = {
      "id": "1234567890",
      "lastname": "yepes",
      "name": "juan",
      "dob": null
    };
    let Render;
    let handleSubmitMock = jest.fn();

    act(() => Render = render(<Create submit={ handleSubmitMock }></Create>));

    await waitFor(() => {
      const inputId = Render.container.querySelector('[name="id"]');
      fireEvent.change(inputId, { target: { value: createDataMock.id } })

      expect(inputId.value).toBe(createDataMock.id);

      const inputName = Render.container.querySelector('[name="name"]');
      fireEvent.change(inputName, { target: { value: createDataMock.name } })

      expect(inputName.value).toBe(createDataMock.name);

      const inputLastName = Render.container.querySelector('[name="lastname"]');
      fireEvent.change(inputLastName, { target: { value: createDataMock.lastname } })

      expect(inputLastName.value).toBe(createDataMock.lastname);

      const inputDob = Render.container.querySelector('[placeholder="DoB"]');
      fireEvent.change(inputDob, { target: { value: "1992-16-02" } })

      expect(inputDob.value).toBe("1992-16-02");

      utils.transformDate = jest.fn(() => null)

      fireEvent.click(Render.container.querySelector('[role="button"]'));

      expect(handleSubmitMock).toBeCalledWith(createDataMock);
    });
  });
});
