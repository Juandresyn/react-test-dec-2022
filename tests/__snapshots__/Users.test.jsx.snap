// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <Users /> component Should call endpoint 1`] = `
<DocumentFragment>
  <div
    class="Users"
  >
    <h1
      class="title__main"
    >
      Manage Users
    </h1>
    <button
      class="button--wide"
      role="button"
    >
      Create User
    </button>
    0 users found
    <table
      class="table"
    >
      <thead>
        <tr>
          <th>
            ID
          </th>
          <th>
            CEDULA
          </th>
          <th>
            NAME
          </th>
          <th>
            LASTNAME
          </th>
          <th>
            AGE
          </th>
          <th>
            OPTIONS
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</DocumentFragment>
`;

exports[`Test <Users /> component Should call endpoint 2`] = `
<DocumentFragment>
  <div
    class="Users"
  >
    <h1
      class="title__main"
    >
      Manage Users
    </h1>
    <button
      class="button--wide"
      role="button"
    >
      Create User
    </button>
    2 users found
    <table
      class="table"
    >
      <thead>
        <tr>
          <th>
            ID
          </th>
          <th>
            CEDULA
          </th>
          <th>
            NAME
          </th>
          <th>
            LASTNAME
          </th>
          <th>
            AGE
          </th>
          <th>
            OPTIONS
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td />
          <td>
            23456789
          </td>
          <td>
            Pepito
          </td>
          <td />
          <td>
            49
          </td>
          <td>
            <button
              id="remove__undefined"
              role="button"
            >
              Remove
            </button>
          </td>
        </tr>
        <tr>
          <td />
          <td>
            12345678
          </td>
          <td>
            Juan
          </td>
          <td />
          <td>
            30
          </td>
          <td>
            <button
              id="remove__undefined"
              role="button"
            >
              Remove
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
